Core: Rollback compaction on conflicts

singhpk234 opened the pull request:
### About the change : 
Revert compaction in case a conflict is detected with the non compaction writes.

* Identify commits  created via compaction, we can use snapshot's `operation()` to check if the given snapshot is of `REPLACE` if yes then it was created as part of compaction.

* While committing the pending updates of the transaction, check if the current snapshot conflicts with the transaction updates. If the current snapshot conflicts and it’s a snapshot created by the compaction process (snapshot#operation()). Then revert the current snapshot (rollback to its parent snapshot) and then try re-applying updates on top of it.
Introduce a new table property `rollback.compaction.on-conflicts.enabled`, which essentially helps in identifying if we want to rollback the compaction commit if conflicts are detected.

* Let’s say a transaction has U1, U2, U3 ... update to be applied, let say B is the base snapshot, now when we were going to actually commit the transaction it saw the current snapshot now is B`(a snapshot created due to compaction), so make current snapshot as B` and try re-apply updates i.e {U1, U2, U3}, when applying U3 on top of { B` -> U1 -> U2 } a metadata conflict was detected, then try updates of transactions by rolling back B` to it’s parent, Let's say rollback B` to its parent is called RollbackB` so now try applying {RollbackB`, U1, U2, U3} on top of B` and see if it still conflicts if yes fail, else commit and update table metadata.

An alternative to proposal : https://docs.google.com/document/d/1pSqxf5A59J062j9VFF5rcCpbW9vdTbBKTmjps80D-B0/edit# suggested in community sync.

Adding this here to get a feedback on the approach. Once agreed upon, can make engines (for ex: spark) use transaction api.

-----

### Testing done : 
- Added ut's for validating is_compacted is present in compacted snapshot summary
- Added ut for e2e using transaction API and concurrent rewrite.
- TODO add more exhaustive UT's 


cc @rdblue @danielcweeks @jackye1995 @rajarshisarkar @amogh-jahagirdar

jackye1995 commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -479,6 +480,52 @@ private void applyUpdates(TableOperations underlyingOps) {
+
+            while (rollbackToSnapshotId != null
+                && !updatesAppliedSuccessfully
+                && PropertyUtil.propertyAsBoolean(
```
I discussed with you a bit offline as well, I think instead of introducing a new property of `is-compacted`, it seems like we can rollback whenever we have conflict and `current.snapshot(currentSnapshotId).equals(DataOperations.REPLACE)`? Which applies to both compaction and rewrite manifests and feels more generic. Is there any risk of doing that way?

singhpk234 replied:
>  I discussed with you a bit offline as well, I think instead of introducing a new property of is-compacted, it seems like we can rollback whenever we have conflict and current.snapshot(currentSnapshotId).equals(DataOperations.REPLACE)

Apologies, I missed that. I think it makes sense to use snapshot#operation to identify if it's a compaction. Updated the PR & description.

jackye1995 commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -479,6 +480,49 @@ private void applyUpdates(TableOperations underlyingOps) {
+          if (PropertyUtil.propertyAsBoolean(
+              base.properties(),
+              TableProperties.ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED,
+              TableProperties.ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED_DEFAULT)) {
```
thanks for the update! In that case should we call this something like `ROLLBACK_REWRITE_ON_CONFLICTS_ENABLED`?

jackye1995 replied:
also the term rollback seems to be not accurate. We are not just rolling back, but we are also reapplying the changes after rolling back. Maybe something like `REAPPLY_REPLACE_SNAPSHOT_ON_CONFLICTS_ENABLED`?

singhpk234 replied:
> We are not just rolling back, but we are also reapplying the changes after rolling back. Maybe something like REAPPLY_REPLACE_ON_CONFLICTS_ENABLED? Any other suggestions for the name?

makes sense, how about something like `RETRY_WITH_ROLLBACK_REPLACE_ON_CONFLICTS_ENABLED`, it gives an impression that we are retrying the applying the changes with roll-backing replace operations snapshot.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -466,6 +467,7 @@ private void cleanUpOnCommitFailure() {
         .run(ops.io()::deleteFile);
   }
 
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
Rather than supressing this, it's a good sign that the new code should be refactored.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -479,6 +481,55 @@ private void applyUpdates(TableOperations underlyingOps) {
+                && !updatesAppliedSuccessfully
+                && DataOperations.REPLACE.equals(current.snapshot(currentSnapshotId).operation())) {
+              SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(transactionOps);
+              setSnapshotOp.rollbackTo(rollbackToSnapshotId);
```
Doesn't this need to be committed?

rdblue reviewed (changes requested):
I think that we need to change the approach in this change. The current PR updates transactions, but that is a harder problem because there are multiple commits to handle. It also performs a somewhat blind rewrite by just checking to see if rolling back works, rather than going back to a starting snapshot. That seems really risky to me.

I think that this work should be done in `SnapshotProducer` rather than in `BaseTransaction`. That will work for direct commits as well as transactions and will be easier to think through. This should also ensure that if there is a starting snapshot set, that the starting snapshot is the furthest that the producer will roll back to.

singhpk234 commented:
Apologies for the delay in updating the pr, i was in the middle of changing my work location.

Have addressed the feedback and moved the changes to SnapshotProducer from BaseTransaction.

@rdblue @jackye1995 please have a look when you get some time.

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,45 @@ private static void updateTotal(
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
+    Long newParentSnapshotId = parentSnapshotId;
+    if (PropertyUtil.propertyAsBoolean(
```
can we move this property check outside this method? So we don't have such a long if block.

jackye1995 commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -284,6 +284,10 @@ private TableProperties() {}
   public static final String WRITE_AUDIT_PUBLISH_ENABLED = "write.wap.enabled";
   public static final String WRITE_AUDIT_PUBLISH_ENABLED_DEFAULT = "false";
 
+  public static final String ROLLBACK_COMPACTION_ON_CONFLICTS_ENABLED =
```
I think the name feels a bit awkward to me, first it's a write side feature, so I think it's better to make it start with `write.`. Secondly this is to rollback a REPLACE operation, so feels to me it's better to call it something like `write.rollback-rewrite-enabled`?

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -215,7 +219,18 @@ public Snapshot apply() {
+      validate(base, parentSnapshot);
+    } catch (ValidationException ve) {
+      parentSnapshotId = snapshotIdToRollbackToOnConflict(parentSnapshotId);
+      // if no snapshot is found, so that roll backing to it avoids conflicts,
```
this sentence does not make much sense to me... can you rephrase?

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
+    }
+  }
+
+  private Boolean shouldRollbackReplaceOnConflict() {
```
it already has a default, can be type `boolean`

jackye1995 commented on core/src/main/java/org/apache/iceberg/BaseTransaction.java:
```diff
@@ -73,6 +73,7 @@ enum TransactionType {
   private TableMetadata current;
   private boolean hasLastOpCommitted;
   private final MetricsReporter reporter;
+  private boolean shouldRollbackCompaction;
```
this name should be updated based on the config name

jackye1995 replied:
and seems like it's not used?

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -215,7 +219,18 @@ public Snapshot apply() {
+    try {
+      validate(base, parentSnapshot);
+    } catch (ValidationException ve) {
+      parentSnapshotId = snapshotIdToRollbackToOnConflict(parentSnapshotId);
```
I think the logic that makes more sense in my mind is:

```
try {
  validate(base, parentSnapshot);
} catch (ValidationException ve) {
  if (shouldRollbackReplaceOnConflict()) {
     // do rollback logic 
 } else {
   // rethrow exception
 }
}
```

instead of having the logic to determine the flag as a part of `snapshotIdToRollbackToOnConflict` which is already determining the new parent snapshot.

What do you think?

singhpk234 replied:
sounds good, made the changes

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
+  private Boolean shouldRollbackReplaceOnConflict() {
+    return PropertyUtil.propertyAsBoolean(
+        base.properties(),
+        TableProperties.COMMIT_ROLLBACK_REPLACE_ON_CONFLICT_ENABLED,
```
I am a bit not sure if this should be a table level property. I would imagine this to enabled only on certain sessions, but the majority of writers still do not rollback. Is there any other option to make this a runtime config instead?

jackye1995 replied:
Explored a bit more offline, I think maybe it's better to do this through an update in related action APIs, for example:

```
public interface SupportsRollbackReplace<T> {
  
  T rollbackReplaceOnConflict();
}

public interface AppendFiles extends SnapshotUpdate<AppendFiles>, SupportsRollbackReplace<AppendFiles> {
  ...
}
```

So that in core we can do:

```
table.newAppend().rollbackReplaceOnConflict().addFiles(...).commit()
```

And then we can make this a session config or runtime option in engine.

What do you think?

jackye1995 replied:
we could potentially make this more generic, for example just add this to `SnapshotUpdate<T>`

```
  T rollback(List<String> operations);
```

Although I don't know what would be the use case of the other kinds of rollback, but seems like a good generalization.

@rdblue do you have any thoughts and further comments regarding this PR?

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
+    if (shouldRollbackReplaceOnConflict()) {
+      // add a set snapshot op on top of base to roll back to parent snapshot
+      boolean isCommitSuccessfullyApplied = false;
+      // Update parentSnapshot to it's grandParent
```
you mean ancestor?

singhpk234 replied:
yup

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
+          validate(tempTableOps.current(), parentSnapshot);
+          isCommitSuccessfullyApplied = true;
+        } catch (ValidationException validationException) {
+          // swallow the exception for re-trying
```
can we add a log message here, something like "Rollback to %s and commit still failed, ...."

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
+        Snapshot parentSnapshot = base.snapshot(newParentSnapshotId);
+
+        SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(tempTableOps);
+        setSnapshotOp.rollbackTo(newParentSnapshotId).commit();
```
this is not within a transaction, wouldn't this actually commit the table?

singhpk234 replied:
As per my understanding, since this is based on `TableOperations tempTableOps = ops.temp(base);` so it should not commit, we use this for validation to create a new temp metadata which will act as an uncommitted table metadata similar to what's done in transaction.

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -615,4 +633,47 @@ private static void updateTotal(
     }
   }
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
```
can we add a javdoc for this to explain the algorithm instead of embedding all the comments inline? Also the method name feels like just getting the ID, but in fact you are already performing the rollback. So maybe an actionable method name like just `rollbackReplaceOnConflictToParent(Long originalParent)`?

singhpk234 replied:
we are not actually doing a rollback here, but rather trying to see which parent snapshot id we can rollback to, once we figure that out we apply the current on this parent snapshot id we found and commit to the table so the actual rollback is happening in `apply()` hence named this method as it is, please let me know your thoughts, happy to make the changes accordingly

jackye1995 commented on core/src/main/java/org/apache/iceberg/SnapshotProducer.java:
```diff
@@ -199,6 +201,8 @@ protected void validate(TableMetadata currentMetadata, Snapshot snapshot) {}
    */
   protected abstract List<ManifestFile> apply(TableMetadata metadataToUpdate, Snapshot snapshot);
 
+  protected abstract Long startingSnapshotId();
```
I think this has to be non null? For the cases which it is currently null, it's just because the value is missing. But we always know what is the starting snapshot ID of the commit.

singhpk234 replied:
are you suggesting the `starting snapshot ID` will be the base on which we are applying the snapshot ?
